### DISCUSSION TITLE
tutorials/lift-off-part2/07-connecting-the-dots-in-server-land

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#32] [course feedback]
# tutorials/lift-off-part2/07-connecting-the-dots-in-server-land

https://www.apollographql.com/tutorials/lift-off-part2/07-connecting-the-dots-in-server-land

**Comment by aniketicloud** [comment]
- Why are we having the context as async function ?
- We are not waiting inside the context function.

**Reply by mabuyo** [reply]
Hey @aniketicloud, great question! `context` needs to be an async function [according to the documentation for Apollo Server](https://www.apollographql.com/docs/apollo-server/data/context/#the-context-function). This is also helpful for establishign database connections and waiting for other operations to complete.

**Comment by vasily-mishanin** [comment]
I'm not a JS developer - I am permanently-new-boilerplate-learner 
Graph QL is awesome, only need to wrap my head around it

**Reply by belloshehu** [reply]
It is truly awesome! Just the steep learning curve that one has to try to work hard to overcome before starting to reap the enormous reward in using GraphQL.

**Comment by Hydrock** [comment]
We are throw cache into TrackAPI :  "new TrackAPI({ cache })"

why?

Our TrackAPI class doesnt hava a constructor to work with it.

It's just misleading.

Yes, I understand. Sometimes it is impossible to make all explanations consistently. But firstly, it seems that this is not the case, and secondly, if it is really necessary, put notes. For example:

"""
Yes, our TrackAPI class does not yet have a constructor to handle the cache parameter. But don't worry, we will definitely process this argument. This is just a reserve for the future.
"""

**Reply by vicenteguedes** [reply]
It extends from `RESTDataSource`, which does have that constructor to work with cache.

**Comment by neogeek13** [comment]
For those of us who selected Typescript, the file in the repo is index.js. I grabbed my index.ts from the first course and created an index.ts and deleted the index.js file.

**Reply by neogeek13** [reply]
Same with schema.js/ts - was getting an error on the import entry for TypeDef until I did so.

**Reply by neogeek13** [reply]
Note that the final folder contains the same mix of .ts and .js. Pretty confusing if we're trying to keep consistent to TS throughout.

**Reply by mabuyo** [reply]
Hey @neogeek13 - can you confirm you cloned the TypeScript version of the repo? The command was `git clone -b typescript https://github.com/apollographql/odyssey-lift-off-part2` - it's the same repo URL for both languages, but the TS version is under the `typescript` branch, [which should have the TS files.](https://github.com/apollographql/odyssey-lift-off-part2/tree/typescript/server/src)

**Reply by mabuyo** [reply]
I'm not sure where you're seeing the mix of .ts and .js files.

Please also note that the _video_ only shows the JavaScript version of the course (though it should be fairly similar).

**Reply by neogeek13** [reply]
Yup - that was the problem. When I went into Course 2 the first lesson didn't have any code so I didn't realize I was on the JS instructions untile lesson 2 at which point I switched. Didn't think about there being a different branch for TS but makes sense. Maybe a callout before the clone statement:

Clone the repository
First, be certain you have selected your preferred language (JavaScript or TypeScript) for this course.

In the directory of your choice with your preferred terminal, clone the app's starter repository:

**Reply by mabuyo** [reply]
Ahh yes, that's it! Thanks for catching that, we'll add some clarifying messages shortly!

**Comment by tomwang57** [comment]
How does the server know it need to call the `Track` resolver? What if there is not such a resolver called `Track`?

**Reply by lizhennessy** [reply]
Hi there @tomwang57! The server calls resolver functions based on which fields are requested in a GraphQL operation. In the next lesson, we'll run a live query against our server, and this is where all the pieces come together: when we choose a field on our schema's `Query` type to query, the schema has the information about what type it returns. 

For instance, `tracksForHome` returns a non-null list of non-null `Track` objects.

```
  type Query {
    "Query to get tracks array for the homepage grid"
    tracksForHome: [Track!]!
     // other Query entrypoints
}
```

To resolve a query for this field in particular, the `Query.tracksForHome` resolver function is called. It's our job to write the function so that it returns a list of objects that match our `Track` definition. When it comes to querying specific fields on the `Track` type, this is where our `Track` resolvers come in handy. I'd recommend jumping back to the [Implementing query resolvers](https://www.apollographql.com/tutorials/lift-off-part2/06-implementing-query-resolvers) lesson, which goes much more in-depth on how this works altogether! But please don't hesitate to reach out if you have any other questions.

**Comment by rickyyc** [comment]
Hello,

I am relatively new to TypeScript ( more used to compiled languages with full-fledged IDEs ), and I am getting this error below. What gives? I ended up copying and pasting from the solution but the error persists. Could anyone help?
 
`
src/index.ts(9,25): error TS2769: No overload matches this call.
  Overload 1 of 2, '(server: ApolloServer<BaseContext>, options?: StartStandaloneServerOptions<BaseContext> & { listen?: ListenOptions; }): Promise<...>', gave the following error.
    Argument of type 'ApolloServer<{ datasources: any; }>' is not assignable to parameter of type 'ApolloServer<BaseContext>'.
      Property 'datasources' is missing in type 'BaseContext' but required in type '{ datasources: any; }'.
  Overload 2 of 2, '(server: ApolloServer<{ datasources: any; }>, options: StartStandaloneServerOptions<{ datasources: any; }> & Required<Pick<StartStandaloneServerOptions<{ datasources: any; }>, "context">> & { ...; }): Promise<...>', gave the following error.
    Type '() => Promise<{ dataSources: { trackAPI: TrackAPI; }; }>' is not assignable to type 'ContextFunction<[StandaloneServerContextFunctionArgument], { datasources: any; }>'.
      Type 'Promise<{ dataSources: { trackAPI: TrackAPI; }; }>' is not assignable to type 'Promise<{ datasources: any; }>'.
        Property 'datasources' is missing in type '{ dataSources: { trackAPI: TrackAPI; }; }' but required in type '{ datasources: any; }'.
`

**Reply by lizhennessy** [reply]
Hey there @rickyyc ! I'm happy to take a look - would you mind pasting what you currently have in your `index.ts` file, or linking the repo where your code is pushed?

**Reply by rickyyc** [reply]
Hi @lizhennessy,

Below is the code which I literally copied and pasted from the course. Thanks!
```
import { ApolloServer } from "@apollo/server";
import { startStandaloneServer } from "@apollo/server/standalone";
import { typeDefs } from "./schema";
import { resolvers } from "./resolvers";
import { TrackAPI } from "./datasources/track-api";

async function startApolloServer() {
  const server = new ApolloServer({ typeDefs, resolvers });
  const { url } = await startStandaloneServer(server, {
    context: async () => {
      const { cache } = server;
      return {
        dataSources: {
          trackAPI: new TrackAPI({ cache }),
        },
      };
    },
  });
  console.log(`
    🚀  Server is running!
    📭  Query at ${url}
  `);
}

startApolloServer();
```

**Reply by BlenderDude** [reply]
@rickyyc, Typescript errors can be cryptic and difficult to decipher. I find reading them from the bottom up is sometimes the most effective. Looking at the very last line of the error:

`Property 'datasources' is missing in type '{ dataSources: { trackAPI: TrackAPI; }; }' but required in type '{ datasources: any; }'.`

It notes that it needs a property `datasources` but is missing in your type `{ dataSources: { trackAPI: TrackAPI; }; }`; looks like it boils down to a capitalization error. `dataSources` -> `datasources` in your `startStandaloneServer` call should fix the issue!

**Reply by rickyyc** [reply]
@BlenderDude Hello, thank you for replying. I managed to fix the error earlier. 

"I find reading them from the bottom up is sometimes the most effective." is a very good tip. Thank you

**Comment by vikyw89** [comment]
the typehint from TS types are missleading

**Reply by lizhennessy** [reply]
Hey there @vikyw89, thank you for the feedback! Are these the type hints that appear in your code editor, or a particular part of the lesson that you feel could be improved? We'd love to make this a better experience for future learners!

**Comment by ced-mos** [comment]
Hey,
in the Code Challenge requirement to add a cache is missing, although it's afterwards tested. Hence, for better DX and consistency reasons it would be nice to add this in the notes as the rest of the tutorial is already perfectly aligned.
Thank you for your effort.

**Reply by mabuyo** [reply]
Thank you for the feedback @ced-mos!

**Comment by belloshehu** [comment]
This chapter clears the confusion from the previous one. No wonder it was titled "Connecting the dots in server-land"!

**Comment by loginov-rocks** [comment]
While it's a common practice to introduce caching solutions to speed up web applications, it's critical to understand that caching can have a significant impact on a system's architecture and isn't something that should be blindly implemented. As a suggestion, it might be better not to recommend cache by default.

Caching, although beneficial at times, isn't needed in every scenario and it can introduce additional complexity to a system. It's a powerful tool that can reduce latency and take load off a system, but it can also cause stale data issues, increase memory consumption, and become a potential point of failure.

Pushing beginners to configure their GraphQL servers with caches without proper understanding can be a bad idea. Beginners, who may not have a deep understanding of server infrastructure or caching mechanisms, might build applications that use their servers primarily as caching servers, which is an unintended and potentially damaging practice.

Without thoroughly understanding how caches operate, they can use up significant resources, and improperly invalidated caches can serve outdated data. Readers learning to use caching prematurely might end up unintentionally hampering the scalability and performance of their servers.

It would therefore be much more appropriate for the course authors to create separate articles or lessons discussing cache mechanisms in depth, when and how to appropriately use them, and the potential impacts of improper caching. This would cater for learners who are interested in caching, and would help to prevent beginners from misusing it.

**Comment by mitesh1103** [comment]
I am getting the following error:

/Users/c083443/Documents/GraphQL_Training/odyssey-lift-off-part2/server/node_modules/@graphql-tools/schema/cjs/addResolversToSchema.js:65
                            throw new Error(`${typeName}.${fieldName} defined in resolvers, but not in schema`);
                                  ^

Error: Query.Track defined in resolvers, but not in schema
    at addResolversToSchema (/Users/c083443/Documents/GraphQL_Training/odyssey-lift-off-part2/server/node_modules/@graphql-tools/schema/cjs/addResolversToSchema.js:65:35)


My resolvers.js file:

const resolvers = {
    Query: {
        // returns an array of Tracks that will be used to populate
        // the homepage grid of our web client
        tracksForHome: (_, __, { dataSources }) => {
            return dataSources.trackAPI.getTracksForHome(); 
          },
          Track: {
            author: ({authorId}, _, {dataSources}) => {
                return dataSources.trackAPI.getAuthor(authorId);
            },
          },
    }
};
module.exports = resolvers;


my schema.js file:

const gql = require('graphql-tag');

const typeDefs = gql`
  type Query {
    "Query to get tracks array for the homepage grid"
    tracksForHome: [Track!]!
  }

  "A track is a group of Modules that teaches about a specific topic"
  type Track {
    id: ID!
    "The track's title"
    title: String!
    "The track's main Author"
    author: Author!
    "The track's illustration to display in track card or track page detail"
    thumbnail: String
    "The track's approximate length to complete, in minutes"
    length: Int
    "The number of modules this track contains"
    modulesCount: Int
  }

  "Author of a complete Track or a Module"
  type Author {
    id: ID!
    "Author's first and last name"
    name: String!
    "Author's profile picture"
    photo: String
  }
`;

module.exports = typeDefs;

**Reply by lizhennessy** [reply]
Hi there @mitesh1103! It looks like the error is in your `resolvers` object; `Query` and `Track` should be two entries in the object, but it looks like `Track` is nested under `Query`.

```
const resolvers = {
    Query: {
         // returns an array of Tracks that will be used to populate
         // the homepage grid of our web client
         tracksForHome: (_, __, { dataSources }) => {
             return dataSources.trackAPI.getTracksForHome();
         },
    },
    Track: {
        author: ({authorId}, _, {dataSources}) => {
             return dataSources.trackAPI.getAuthor(authorId);
        },
    },
};
module.exports = resolvers;
```

Moving the closing bracket of the `Query` object (`}`) up should fix it!

**Comment by samuelpucat** [comment]
For curious cats, like myself, wondering whether it's possible to use both mocks and data from resolvers - yes you can:
1. Define your new, not yet implemented entrypoint to schema.ts:
```ts
type Query {
  ...
  "not yet implemented"
  testingMock: Int!
}
```
2. Add mocks for this entrypoint:
```ts
const mocks = {
  Query: () => ({
    ...
    testingMock: () => 123456,
  }),
  ...
};
```
3. Define ApolloServer like so:
```ts
const server = new ApolloServer({
  schema: addMocksToSchema({
    schema: makeExecutableSchema({ typeDefs, resolvers }),
    mocks,
    preserveResolvers: true,
  }),
});
```
Source: https://www.apollographql.com/docs/apollo-server/testing/mocking

**Comment by rvelasquez** [comment]
Have a strange issue when running the server.

Error: "tracksForHome" defined in resolvers, but has invalid value "(_, __, { dataSources }) => {
        return dataSources.trackAPI.getTracksForHome();
    }". The resolver's value must be of type object.

I've checked the schema:

  type Query {
    "Get tracks array for homepage grid"
    tracksForHome: [Track!]!
  }

and the resolver:

Query: {
        tracksForHome: (_, __, { dataSources }) => {
            return dataSources.trackAPI.getTracksForHome();
        },
    },

What am I doing wrong?

**Reply by rvelasquez** [reply]
nvm, user error.  I forgot to save a file in the IDE so it wasn't loading the right code.

**Comment by srinivas-sampath** [comment]
Hi Folks, i was going thro' the   lift-off-part2 of the odyssey course and the section where it is talking about "connecting the dots in server-land" where we launch the sandbox explorer, when i try to invoke my existing query, it keeps giving "INTERNAL SERVER ERROR".
Error returned is this
{
  "errors": [
    {
      "message": "Context creation failed: TrackAPI is not a constructor",
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "stacktrace": [
          "TypeError: TrackAPI is not a constructor",
          "    at context (/Users/srinivassampath/odyssey/odyssey-lift-off-part2/server/src/index.js:15:21)",
          "    at context (/Users/srinivassampath/odyssey/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/express4/index.js:32:28)",
          "    at ApolloServer.executeHTTPGraphQLRequest (/Users/srinivassampath/odyssey/odyssey-lift-off-part2/server/node_modules/@apollo/server/dist/cjs/ApolloServer.js:531:38)",
          "    at process.processTicksAndRejections (node:internal/process/task_queues:104:5)"
        ]
      }
    }
  ]
}

The funny thing is that i even copy pasted the code from the course "verbatim" and i am still getting the same error.  Not sure what i am doing wrong.  Any help is appreciated.  Thanks

**Reply by srinivas-sampath** [reply]
Issue is resolved.  Looks like i did not save the track-api.js after making changes and hence the file was out-of-date.
